File: src/main/java/com/riojasonc/sphere/servlet/Hitokoto.java
package com.riojasonc.sphere.servlet;

import net.sf.json.JSONObject;

import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.io.PrintWriter;
import java.sql.SQLException;

import static com.riojasonc.sphere.lib.BasicFunctions.flushErrorStatus;
import static com.riojasonc.sphere.lib.BasicFunctions.getJSONFromServletRequest;
import static com.riojasonc.sphere.lib.util.Database.getHitokoto;

@WebServlet(name = "Hitokoto", urlPatterns = "/function/hitokoto/get")
public class Hitokoto extends HttpServlet {
    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) {

    }

    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
        response.setCharacterEncoding("UTF-8");
        response.setContentType("application/json;charset=UTF-8");

        JSONObject requestJson = getJSONFromServletRequest(request);
        if(requestJson == null) {
            return;
        }
        JSONObject responseJson = new JSONObject();
        JSONObject innerRequestJson = new JSONObject();
        PrintWriter pw = response.getWriter();

        if("getMaxId".equals(requestJson.getString("mode"))) {
            innerRequestJson.put("mode", "getMaxId");
            JSONObject innerResponseJson;
            try {
                innerResponseJson = getHitokoto(innerRequestJson);
                long id = innerResponseJson.getLong("maxId");
                if(id == -1) {
                    responseJson.put("Status", "Error");
                    responseJson.put("ErrorType", "Server Internal Error");
                }
                else {
                    responseJson.put("Status", "Success");
                    responseJson.put("data", id);
                }
                pw.print(responseJson);
                pw.flush();
                pw.close();;
            }
            catch (SQLException e) {
                flushErrorStatus(pw, e.getMessage());
                e.printStackTrace();
            }
        }
        else {
            innerRequestJson.put("mode", "getHitokoto");
            innerRequestJson.put("id", requestJson.getLong("id"));
            JSONObject innerResponseJson;
            try {
                innerResponseJson = getHitokoto(innerRequestJson);
                String hitokoto = innerResponseJson.getString("hitokoto");
                if("Error".equals(hitokoto)) {
                    responseJson.put("Status", "Error");
                    responseJson.put("ErrorType", "Server Internal Error");
                }
                else {
                    responseJson.put("Status", "Success");
                    responseJson.put("hitokoto", hitokoto);
                    responseJson.put("from_who", innerResponseJson.getString("from_who"));
                    responseJson.put("from", innerResponseJson.getString("hfrom"));
                }
                pw.print(responseJson);
                pw.flush();
                pw.close();;
            }
            catch (SQLException e) {
                flushErrorStatus(pw, e.getMessage());
                e.printStackTrace();
            }
        }
    }
}

File: src/main/webapp/js/hitokoto/hitokoto.js
let myHitokotoMaxId = -1;
let myHitokotoScale = -1;

getMyHitokotoMaxId();
setTimeout(function () {
    myHitokotoScale = myHitokotoMaxId / (54.0 + myHitokotoMaxId);
}, 200);

function getHitokoto() {
    let result = {Status: "Success"};
    $.ajax({
        url: "https://v1.hitokoto.cn/",
        type: "post",
        data: JSON.stringify({
            encode: "json",
            charset: "utf8"
        }),
        dataType: "json",
        success: function (response) {
            if(response["from_who"] === "null")
                response["from_who"] = null;
            if(response["from"] === "null")
                response["from"] = null;
            $.extend(result, response);
        },
        error: function () {
            result = {Status: "Error"};
        }
    });
    return result;
}

function getMyHitokoto(id) {
    let result = {"Status": "Success"};
    $.ajax({
        url: rootUrl + "/function/hitokoto/get",
        type: "post",
        data: JSON.stringify({
            mode: "getHitokoto",
            id: id
        }),
        dataType: "json",
        success: function (response) {
            $.extend(result, response);
        },
        error: function () {
            result = {Status: "Error", ErrorType: "Server Internal Error"};
        }
    });
    return result;
}

function getMyHitokotoMaxId() {
    $.ajax({
        url: rootUrl + "/function/hitokoto/get",
        type: "post",
        data: JSON.stringify({
            mode: "getMaxId"
        }),
        dataType: "json",
        success: function (response) {
            myHitokotoMaxId = response["data"];
        },
        error: function () {
            myHitokotoMaxId = -1;
        }
    });
}